src/Microsoft.AspNetCore.SystemWebAdapters.SessionState/Serialization/JsonSessionSerializer.Shared.cs

```diff
@@ -2,123 +2,56 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
-using System.Collections.Generic;
 using System.Text.Json;
-using System.Text.Json.Serialization;
 
 namespace Microsoft.AspNetCore.SystemWebAdapters.SessionState.Serialization;
 
 internal partial class JsonSessionSerializer : ISessionSerializer
 {
-    private readonly IDictionary<string, Type> _map;
-    private readonly JsonSerializerOptions _options;
+    private readonly JsonSessionSerializerOptions _options;
+    private readonly JsonSerializerOptions _jsonOptions;
 
-    public JsonSessionSerializer(IDictionary<string, Type> map, bool writeIndented = false)
+    private static JsonSerializerOptions BuildJsonOptions(JsonSessionSerializerOptions options) => new()
     {
-        _map = map;
-        _options = new JsonSerializerOptions
-        {
 #if !NETCOREAPP3_1
-            DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingDefault,
+        DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingDefault,
 #endif
-            AllowTrailingCommas = true,
-            WriteIndented = writeIndented,
-            Converters =
-            {
-                new SerializedSessionConverter(map),
-            }
-        };
-    }
-
-    public byte[] Serialize(string key, object value) => _map.TryGetValue(key, out var type)
-        ? JsonSerializer.SerializeToUtf8Bytes(value, type, _options)
-        : throw new InvalidOperationException($"Key '{key}' is not registered");
+        AllowTrailingCommas = true,
+        WriteIndented = options.Indented,
+        Converters =
+        {
+            new SerializedSessionConverter(options.KnownKeys),
+        }
+    };
 
-    public object? Deserialize(string key, Memory<byte> bytes)
+    public byte[] Serialize(string key, object value)
     {
-        if (_map.TryGetValue(key, out var type))
+        if (_options.KnownKeys.TryGetValue(key, out var type))
         {
-            if (bytes.IsEmpty)
-            {
-                return null;
-            }
-
-            return JsonSerializer.Deserialize(bytes.Span, type, _options);
+            return JsonSerializer.SerializeToUtf8Bytes(value, type, _jsonOptions);
         }
 
-        throw new InvalidOperationException($"Key '{key}' is not registered");
+        LogUnknownKey(key, serialize: true);
+
+        return _options.ThrowOnUnknownSessionKey ? throw new UnknownSessionKeyException(new[] { key }) : Array.Empty<byte>();
```

Comment:
Why not always throw?

Reply:
I could imagine where someone just wants to log the missing keys so that the system doesn't crash if a new session key is added. However, the flag does allow someone to opt into it.

As an implementation detail, allowing it to be a flag allows the remote app to not throw and pass the info along, while the core app can then decide to throw or not but uses the same implementation.

Reply:
I would expect the serialization exception to be opt-out. Exceptions are much more discoverable than logs.

Reply:
Oh that's reasonable. I've swapped the default

Reply:
This looks like a good approach. I can definitely imagine someone wanting to opt out if they have a lot of session items but only expect a few of them to be needed by ASP.NET Core endpoints. I like throwing by default, though.